File: apps/pms-api/app/models/platform_admin.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


def to_camel(string: str) -> str:
    parts = string.split("_")
    return parts[0] + "".join(word.capitalize() for word in parts[1:])


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


PropertyStatus = Literal["live", "demo", "test"]
Granularity = Literal["daily", "weekly", "monthly"]


class PlatformProperty(CamelModel):
    id: str
    name: str
    slug: str
    status: PropertyStatus
    created_at: str


class MetricDelta(CamelModel):
    value: float | None
    label: str


class GrowthMetric(CamelModel):
    key: str
    label: str
    value: str
    raw_value: float | int | None
    delta: MetricDelta | None = None


class ChartPoint(CamelModel):
    key: str
    label: str
    value: int


class GrowthDashboardResponse(CamelModel):
    properties: list[PlatformProperty]
    selected_property_ids: list[str]
    exclude_test_data: bool
    granularity: Granularity
    booking_property_id: str | None = None
    metrics: list[GrowthMetric]
    page_views: list[ChartPoint]
    booking_requests: list[ChartPoint]
    live_properties: list[ChartPoint]
    empty_message: str | None = None


class UpdatePlatformPropertyStatusRequest(CamelModel):
    status: PropertyStatus = Field(..., description="Platform classification for analytics")

File: apps/platform-admin/services/platformAdmin.ts
import { pmsApi } from "./api/client";

export type PlatformStatus = "live" | "demo" | "test";
export type Granularity = "daily" | "weekly" | "monthly";

export interface PlatformProperty {
  id: string;
  name: string;
  slug: string;
  status: PlatformStatus;
  createdAt: string;
}

export interface GrowthMetric {
  key: string;
  label: string;
  value: string;
  rawValue: number | null;
  delta: { value: number | null; label: string } | null;
}

export interface ChartPoint {
  key: string;
  label: string;
  value: number;
}

export interface GrowthDashboard {
  properties: PlatformProperty[];
  selectedPropertyIds: string[];
  excludeTestData: boolean;
  granularity: Granularity;
  bookingPropertyId: string | null;
  metrics: GrowthMetric[];
  pageViews: ChartPoint[];
  bookingRequests: ChartPoint[];
  liveProperties: ChartPoint[];
  emptyMessage: string | null;
}

export function getGrowthDashboard(params: {
  granularity: Granularity;
  excludeTestData: boolean;
  propertyIds?: string[];
  bookingPropertyId?: string;
}) {
  const search = new URLSearchParams({
    granularity: params.granularity,
    exclude_test_data: String(params.excludeTestData),
  });
  if (params.propertyIds) {
    if (params.propertyIds.length === 0) {
      search.append("property_ids", "");
    } else {
      params.propertyIds.forEach((id) => search.append("property_ids", id));
    }
  }
  if (params.bookingPropertyId) {
    search.set("booking_property_id", params.bookingPropertyId);
  }
  return pmsApi.get<GrowthDashboard>(`/platform-admin/growth?${search.toString()}`);
}

export function updatePropertyStatus(id: string, status: PlatformStatus) {
  return pmsApi.patch<PlatformProperty>(`/platform-admin/properties/${id}/status`, { status });
}
